### ISSUE TITLE
Serializing and deserializing `Arc<"">` result in UB

### ISSUE THREAD

**Issue by kdy1** [rkyv/rkyv#288]
Related: https://github.com/swc-project/swc/issues/5325

We serialize and deserialize AST to invoke plugins, and I found that serialization of empty `Atom` (https://github.com/swc-project/swc/blob/31b3cf21707f088d8379a03837edca0bd023d587/crates/swc_atoms/src/lib.rs#L58) result in a garbage value

Related: https://github.com/swc-project/swc/pull/5323

**Comment by djkoloski** [comment]
I'm unable to get a failure with this test, which exercises a relatively complex case similar to yours:
```rust
use std::sync::Arc;

#[derive(Clone, Archive, Serialize, Deserialize)]
#[archive_attr(repr(C), derive(CheckBytes))]
pub struct Structure {
    value: String,
    shared: Vec<Arc<str>>,
}

let shared = Arc::<str>::from("");
let value = Structure {
    value: "hello world!".to_string(),
    shared: vec![shared.clone(), shared.clone(), shared.clone()],
};

let mut serializer = DefaultSerializer::default();
serializer.serialize_value(&value).unwrap();
let buf = serializer.into_serializer().into_inner();

let structure = check_archived_root::<Structure>(buf.as_ref()).unwrap();

assert_eq!(&*structure.value, "hello world!");
for s in structure.shared.iter() {
    assert_eq!(&**s, "");
}
```
Is it possible that the `Arc`s being serialized were also being mutated? During serialization, the value of the `Arc` is read twice and if they differ, they may cause UB.

**Comment by kdy1** [comment]
Hmm...
It was a newtype of `Arc<str>`, with derive.

https://github.com/swc-project/swc/blob/31b3cf21707f088d8379a03837edca0bd023d587/crates/swc_atoms/src/lib.rs#L52-L58

Can this make a difference?

**Comment by djkoloski** [comment]
Got a relatively minimal reproduction for this in wasmer thanks to @nagisa, this should be fixed by 58587fdf527546bedd593c649b889898c3266daf.
